avr/cores/megacommand/MCL/MDSeqTrack.cpp

```diff
@@ -700,75 +785,119 @@ void MDSeqTrack::merge_from_md(uint8_t track_number, MDPattern *pattern) {
+    memcpy(locks, lock_buf, nlock);
+
+    // shift steps
+    step_buf = steps[length - 1];
+    timing_buf = timing[length - 1];
+    memmove(steps + 1, steps, ncopy);
+    steps[0] = step_buf;
+    timing[0] = timing_buf;
+    break;
+  }
+  case DIR_REVERSE: {
+    uint8_t rev_locks[NUM_MD_LOCK_SLOTS];
+    memcpy(rev_locks, locks, sizeof(locks));
+    uint16_t l = 0, r = 0;
+    // reverse steps & locks
+    for (int i = 0; i <= length / 2; ++i) {
+      int j = length - i - 1;
+      if (j < i) {
+        break;
       }
-      break;
-    case DIR_REVERSE:
-      new_pos = length - n - 1;
-      break;
-    }
-
-    for (uint8_t a = 0; a < NUM_MD_LOCKS; a++) {
-      locks[a][new_pos] = temp_data.locks[a][n];
-    }
-    conditional[new_pos] = temp_data.conditional[n];
-    timing[new_pos] = temp_data.timing[n];
-    if (IS_BIT_SET64(temp_data.pattern_mask, n)) {
-      SET_BIT64(pattern_mask, new_pos);
-    }
-    if (IS_BIT_SET64(temp_data.lock_mask, n)) {
-      SET_BIT64(lock_mask, new_pos);
+      uint8_t ni = popcount(steps[i].locks);
+      uint8_t nj = popcount(steps[j].locks);
+      memcpy(locks + l, rev_locks + total_nlock - l - nj, nj);
+      memcpy(locks + total_nlock - r - ni, rev_locks + r, ni);
+      l += nj;
+      r += ni;
+      step_buf = steps[i];
+      steps[i] = steps[j];
+      steps[j] = step_buf;
+      timing_buf = timing[i];
+      timing[i] = timing[j];
+      timing[j] = timing_buf;
     }
+    break;
+  }
   }
 }
 
 void MDSeqTrack::copy_step(uint8_t n, MDSeqStep *step) {
   step->active = true;
+  uint8_t idx = get_lockidx(n);
+  uint8_t lcks = steps[n].locks;
+  uint8_t mask = 1;
   for (uint8_t a = 0; a < NUM_MD_LOCKS; a++) {
-    step->locks[a] = locks[a][n];
+    if (lcks & mask) {
+      step->locks[a] = locks[idx++];
+    } else {
+      step->locks[a] = 0;
+    }
+    mask <<= 1;
   }
-  step->conditional = conditional[n];
+  step->conditional = steps[n].cond_id;
+  step->conditional_plock = steps[n].cond_plock;
   step->timing = timing[n];
-  step->lock_mask = IS_BIT_SET64(lock_mask, n);
-  step->pattern_mask = IS_BIT_SET64(pattern_mask, n);
-  step->slide_mask = IS_BIT_SET64(slide_mask, n);
+  step->pattern_mask = steps[n].trig;
+  step->slide_mask = steps[n].slide;
+  step->locks_enabled = steps[n].locks_enabled;
 }
 
 void MDSeqTrack::paste_step(uint8_t n, MDSeqStep *step) {
+  uint8_t idx = get_lockidx(n);
+  uint8_t nlock = popcount(steps[n].locks);
+  int8_t delta = -nlock;
+  uint8_t new_locks = 0;
+
   for (uint8_t a = 0; a < NUM_MD_LOCKS; a++) {
-    locks[a][n] = step->locks[a];
-  }
-  conditional[n] = step->conditional;
-  timing[n] = step->timing;
-  if (step->lock_mask) {
-    SET_BIT64(lock_mask, n);
-  }
-  if (step->pattern_mask) {
-    SET_BIT64(pattern_mask, n);
+    if (step->locks[a]) {
+      ++delta;
+      new_locks |= 1;
+    }
+    new_locks <<= 1;
   }
-  if (step->slide_mask) {
-    SET_BIT64(slide_mask, n);
+
+  if (delta < 0) { // shrink
```

Comment:
You've added the code for shifting the array, but I think you've missed the actual lock copy/insert?

Reply:
actually, is the array shifting needed at all.

can we just recycle

 bool MDSeqTrack::set_track_locks_i(uint8_t step, uint8_t lockidx,
545                                    uint8_t value) {

to save a few bytes?

Reply:
you're right :D just lay down the lock data at idx+nlock, the space is properly reserved.

Reply:
will do

Reply:
> can we just recycle....

set_track_locks only set one lock at a time, so it means resizing the array up to 8 times. so the question is, do we want to pay the price.

.... or, make set_track_locks_i work for multiple locks at once?

Reply:
I suppose if we implement copying multiple steps (copy page for example) it could have more of an impact.

also, what are the other scenario where we will be setting multiple locks together?
In live record we would need to have some buffer mechanism in CC callback to make use of this.

Furthermore, the majority of the work is the 
```
556     memmove(locks + lock_slot + 1, locks + lock_slot,
557             NUM_MD_LOCK_SLOTS - lock_slot - 1);
```

how many instructions is this? for 256 byte array it might not be that bad afterall.

Reply:
Thinking a little larger...

Eventually, we will need to generalise these routines so that they can be shared with the ExtSeqTrack data structures

Recording notes in realtime would probably benefit from having a multiple insert routine.

Reply:
How do we eval the performance of RLCK? I tried a little bit and was surprised it worked well enough for me as-is, without the circular insertion buffer we discussed.

Reply:
The important performance metric is latency. 

If there is a long wait to insert the next lock, it may get applied at the wrong sequencer position.

It might be sufficient to show that insertion time is much shorter than sequencer resolution.  i.e 
t_insertion < MidiClock.div192_time